File: solutions/0200-number-of-islands/solution.py
class Solution:
    deltas = [(0, 1), (1, 0), (0, -1), (-1, 0)]

    def _in_bounds(self, x: int, y: int, max_row: int, max_col: int) -> bool:
        if x < 0 or x >= max_row:
            return False
        if y < 0 or y >= max_col:
            return False
        return True

    def numIslands(self, grid: List[List[str]]) -> int:

        # traverse the grid, one cell at a time, modifying it in place to mark cells as having been visited
        # each cell can be part of exactly one island so marking it as soon as it's seen is good enough

        max_row = len(grid)
        if max_row <= 0:
            return 0
        max_col = len(grid[0])

        def dfs_and_mark(row: int, col: int) -> None:
            grid[row][col] = 'x'
            for delta in Solution.deltas:
                new_row, new_col = row + delta[0], col + delta[1]
                if self._in_bounds(new_row, new_col, max_row, max_col) and grid[new_row][new_col] == '1':
                    dfs_and_mark(new_row, new_col)

        num_components = 0
        for i in range(max_row):
            for j in range(max_col):
                if grid[i][j] == '1':
                    dfs_and_mark(i, j)
                    num_components += 1
        
        return num_components


        

File: solutions/0200-number-of-islands/solution.java
class Solution {

    private static int maxRow;
    private static int maxCol;

    private static boolean isInBounds(int row, int col) {
        if (row < 0 || col < 0) {
            return false;
        }
        if (row >= maxRow || col >= maxCol) {
            return false;
        }
        return true;
    }
    /**
        Modifies the grid in place. Start from grid[i][j]
     */
    private void bfs(int i, int j, char[][] grid) {
        if (grid[i][j] == '1') {
            grid[i][j] = 'X';
        } else {
            return;
        }

        // neighbor down
        if (isInBounds(i+1, j)) {
            bfs(i+1, j, grid);
        }

        // neighbor up
        if (isInBounds(i-1, j)) {
            bfs(i-1, j, grid);
        }

        // neighbor to right
        if (isInBounds(i, j+1)) {
            bfs(i, j+1, grid);
        }

        // neighbor to left
        if (isInBounds(i, j-1)) {
            bfs(i, j-1, grid);
        }
    }

    public int numIslands(char[][] grid) {
        if (grid.length == 0) {
            return 0;
        }

        int rows = grid.length;
        int columns = grid[0].length;
        this.maxRow = rows;
        this.maxCol = columns;

        int numIslands = 0;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
                if (grid[i][j] == '1') {
                    bfs(i, j, grid);
                    numIslands++;
                }
            }
        }

        return numIslands;
    }
}
